bundles/core/src/main/java/com/adobe/cq/commerce/core/components/client/MagentoGraphqlClient.java

```diff
@@ -160,24 +187,53 @@ private MagentoGraphqlClient(Resource resource, Page page) {
             headers.add(new BasicHeader("Store", storeCode));
         }
 
+        Long previewVersion = null;
         if (launch != null) {
             Calendar liveDate = launch.getLiveDate();
             if (liveDate != null) {
                 TimeZone timeZone = liveDate.getTimeZone();
                 OffsetDateTime offsetDateTime = OffsetDateTime.ofInstant(liveDate.toInstant(), timeZone.toZoneId());
-                Long previewVersion = offsetDateTime.toEpochSecond();
-                headers.add(new BasicHeader("Preview-Version", String.valueOf(previewVersion)));
-
-                // We use POST to ensure that Magento doesn't return a cached response
-                requestOptions.withHttpMethod(HttpMethod.POST);
+                previewVersion = offsetDateTime.toEpochSecond();
+            }
+        } else if (request != null) {
+            Long timewarp = getTimeWarpEpoch(request);
+            if (timewarp != null) {
+                Calendar time = Calendar.getInstance();
+                time.setTimeInMillis(timewarp);
+                if (time.after(Calendar.getInstance())) {
+                    previewVersion = timewarp / 1000; // timewarp is in milliseconds, Magento Preview-Version header is in seconds
+                }
             }
         }
 
+        if (previewVersion != null) {
+            headers.add(new BasicHeader("Preview-Version", String.valueOf(previewVersion)));
+
+            // We use POST to ensure that Magento doesn't return a cached response
+            requestOptions.withHttpMethod(HttpMethod.POST);
+        }
+
         if (!headers.isEmpty()) {
             requestOptions.withHeaders(headers);
         }
     }
 
+    private Long getTimeWarpEpoch(SlingHttpServletRequest request) {
+        String timeWarp = request.getParameter("timewarp");
```

Comment:
Is this the only place we use the request? If so, maybe we can refactor a bit to just get the `timewarp` as a parameter when you create the `MagentoGraphqlClient` instead of the full SlingHttpServletRequest object.

Reply:
Well if one day we need something else from the request, then we wouldn't have to change the `create` method API any more, so I'd prefer that we keep the request, also so that "users" of the method do not have to extract the parameter by themselves.